### PR TITLE
[hotfix] stop preprint deletions [OSF-7938]

### PR DESCRIPTION
**Pull request by alexschiller**
## Purpose
Code preventing preprints from being deleted did not get ported over when osfstorage was moved to addons. Currently only fangorn is preventing deletions of preprint files and bad moves, but fangorn has some known flaws (lazy loaded nested folders are supposed to be caught by the backend)

## Changes
Add back in the code blocking bad preprint actions

## Side effects
None known; someone should check how many orphans were created.

## Ticket
https://openscience.atlassian.net/browse/OSF-7938

## QA Notes
This is a regression of https://github.com/CenterForOpenScience/osf.io/pull/6525. However, all of the fangorn logic is still in place so the things that need thorough testing are the areas where fangorn is known to fail: when a preprint is inside a folder that has not been expanded. Test this situation thoroughly. A light regression of the fangorn blocking of bad actions (e.g., it copies instead of moves when it has a preprint; the button is grayed out when you click a preprint) would not be a bad idea.

### CODE REVIEW

File: addons/osfstorage/models.py [hunk 1]
```diff
@@ -342,6 +352,14 @@ def is_checked_out(self):
                 pass
         return False
 
+    @property
+    def is_preprint_primary(self):
```
Comment (alexschiller) [inline addons/osfstorage/models.py]: The first time I wrote this I didn't like it. Having to put it in again still makes me sad. I'm sure there is a better Django-y way to do this (and the function that lives above it related to checked out files).

File: addons/osfstorage/tests/test_views.py [hunk 2]
```diff
@@ -649,6 +665,30 @@ def test_move_checkedout_file(self):
         )
         assert_equal(res.status_code, 405)
 
+    def test_within_node_move_while_preprint(self):
```
Comment (alexschiller) [inline addons/osfstorage/tests/test_views.py]: These three tests were what I had included originally. There are some more edge cases that would ideally be covered. A followup ticket for that (and the note about is_preprint_primary) is probably needed.

File: addons/osfstorage/models.py (hunk 1 above)
Comment (mfraezz) [inline addons/osfstorage/models.py]: A recursive CTE would work, but I don't know of any Django magic for this. Check out `AbstractNodeQuerySet.get_children` if you're feeling adventurous -- you'd want something similar.

File: addons/osfstorage/tests/test_views.py (hunk 2 above)
Comment (brianjgeiger) [inline addons/osfstorage/tests/test_views.py]: Go ahead and make that. Might as well add an API test to disallow deleting a preprint file as well, since that's a case that also would have been caught if this had disappeared.

File: addons/osfstorage/models.py [hunk 3]
```diff
@@ -128,17 +128,27 @@ def path(self):
...
+        return self.node.preprint_file == self and not self.node._has_abandoned_preprint
+
+    def delete(self, user=None, parent=None, **kwargs):
         self._path = self.path
         self._materialized_path = self.materialized_path
-        return super(OsfStorageFileNode, self).delete(user=user, parent=parent, **kwargs)
+        return super(OsfStorageFileNode, self).delete(user=user, parent=parent) if self._delete_allowed else None
 
     def move_under(self, destination_parent, name=None):
+        if self.is_preprint_primary:
```
Comment (sloria) [inline addons/osfstorage/models.py]: Better to move this below the `is_checked_out` block, so we can avoid this expensive operation for checked out files.

File: addons/osfstorage/models.py (hunk 1 above)
Comment (sloria) [inline addons/osfstorage/models.py]: @alexschiller For now, just make a DjangoLE ticket to investigate optimizing this.

File: addons/osfstorage/models.py [hunk 4]
```diff
@@ -342,6 +352,14 @@ def is_checked_out(self):
                 pass
         return False
 
+    @property
+    def is_preprint_primary(self):
+        if self.node.is_preprint:
+            for child in self.children:
```
Comment (sloria) [inline addons/osfstorage/models.py]: You should be able to do `.select_related('node')`

File: addons/osfstorage/models.py (hunk 3 above)
Comment (alexschiller) [inline addons/osfstorage/models.py]: @sloria Don't worry, checked out files make an equally expensive query... Might be able to combine them. 

https://github.com/alexschiller/osf.io/blob/a202b3c04fcff401e0ada1b3887dc9e6c6a21b6e/addons/osfstorage/models.py#L347

File: addons/osfstorage/models.py [hunk 5]
```diff
@@ -128,17 +128,27 @@ def path(self):
     def is_checked_out(self):
         return self.checkout is not None
 
-    def delete(self, user=None, parent=None, **kwargs):
-        if self.node.preprint_file and self.node.preprint_file.pk == self.pk:
-            self.node._is_preprint_orphan = True
-            self.node.save()
+    @property
+    def _delete_allowed(self):
```
Comment (sloria) [inline addons/osfstorage/models.py]: Minor: It feels a bit unintuitive for a property to do validation and raise validation errors. Perhaps make this a method, e.g. `_check_delete_allowed()`. I don't feel strongly about it though.

File: addons/osfstorage/models.py [hunk 6]
```diff
@@ -128,17 +128,27 @@ def path(self):
...
+
+    def delete(self, user=None, parent=None, **kwargs):
         self._path = self.path
         self._materialized_path = self.materialized_path
-        return super(OsfStorageFileNode, self).delete(user=user, parent=parent, **kwargs)
+        return super(OsfStorageFileNode, self).delete(user=user, parent=parent) if self._delete_allowed else None
 
     def move_under(self, destination_parent, name=None):
+        if self.is_preprint_primary:
+            if self.node._id != destination_parent.node._id or self.provider != destination_parent.provider:
```
Comment (sloria) [inline addons/osfstorage/models.py]: You can avoid some queries by not accessing `_id`:

```python
if self.node != destination_parent.node:

# OR

if self.node.pk != destination_parent.node.pk:
```

File: addons/osfstorage/models.py (hunk 1 above)
Comment (alexschiller) [inline addons/osfstorage/models.py]: @sloria https://openscience.atlassian.net/browse/OSF-7958

File: addons/osfstorage/models.py (hunk 3 above)
Comment (sloria) [inline addons/osfstorage/models.py]: Ah, I see. That would be worth mentioning in the ticket suggested in https://github.com/CenterForOpenScience/osf.io/pull/7194/files?diff=split#r115499174

Comment (sloria) [inline addons/osfstorage/models.py]: Also, minor thing I noticed: `children.all()` is called in the code you linked; is that also necessary for `is_preprint_primary`?

File: addons/osfstorage/models.py (hunk 4 above)
Comment (alexschiller) [inline addons/osfstorage/models.py]: Deferring this change to https://openscience.atlassian.net/browse/OSF-7954 to not hold up this hotfix while waiting for travis.

**Comment by brianjgeiger** [comment]
```
jshint /home/travis/build/CenterForOpenScience/osf.io/website/static/js
/home/travis/build/CenterForOpenScience/osf.io/website/static/js/filepage/index.js: line 514, col 17, Confusing use of '!'.
```
